src/python/alog/protocols.py

```diff
@@ -0,0 +1,576 @@
+
+    def addHandler(self, hdlr: "Handler") -> None:  # pylint: disable=invalid-name
+        """
+        Add the specified handler to this logger.
+        """
+
+    def removeHandler(self, hdlr: "Handler") -> None:  # pylint: disable=invalid-name
+        """
+        Remove the specified handler from this logger.
+        """
+
+    def findCaller(  # pylint: disable=invalid-name
+        self, stack_info: bool = False, stacklevel: int = 1
+    ) -> Tuple[str, int, str, Optional[str]]:  # type: ignore
+        """
+        Find the stack frame of the caller so that we can note the source
+        file name, line number and function name.
+        """
+
+    def handle(self, record: "LogRecord") -> None:
+        """
+        Call the handlers for the specified record.
+
+        This method is used for unpickled records received from a socket, as
+        well as those created locally. Logger-level filtering is applied.
+        """
+
+    def makeRecord(  # pylint: disable=invalid-name
+        self,
+        name: str,
+        level: int,
+        fn: str,
+        lno: int,
+        msg: object,
+        args: "_ArgsType",
+        exc_info: Optional["_SysExcInfoType"],
+        func: Optional[str] = None,
+        extra: Optional[Mapping[str, object]] = None,
+        sinfo: Optional[str] = None,
+    ) -> "LogRecord":  # type: ignore
+        """
+        A factory method which can be overridden in subclasses to create
+        specialized LogRecords.
+        """
+
+    def hasHandlers(self) -> bool:  # type: ignore pylint: disable=invalid-name
+        """
+        See if this logger has any handlers configured.
+
+        Loop through all handlers for this logger and its parents in the
+        logger hierarchy. Return True if a handler was found, else False.
+        Stop searching up the hierarchy whenever a logger with the "propagate"
+        attribute set to zero is found - that will be the last logger which
+        is checked for the existence of handlers.
+        """
+
+    def callHandlers(self, record: "LogRecord") -> None:  # pylint: disable=invalid-name
+        """
+        Pass a record to all relevant handlers.
+
+        Loop through all handlers for this logger and its parents in the
+        logger hierarchy. If no handler was found, output a one-off error
+        message to sys.stderr. Stop searching up the hierarchy whenever a
+        logger with the "propagate" attribute set to zero is found - that
+        will be the last logger whose handlers are called.
+        """
+
+
+class ALogLoggerProtocol(LoggerProtocol):
+    """
+    A Protocol which describes the alchemy logging methods which it adds onto the
+    stdlib `logging.Logger` instances.
+    """
+
+    def isEnabled(self, level: "_Level") -> bool:  # type: ignore pylint: disable=invalid-name
+        """
+        Is this logger enabled for level 'level'?
+
+        Also supports referencing the level by name.
+        """
+
+    def off(
```

Comment:
I don't think we want this function. The `"off"` level only exists as a valid value for configuration to fully disable the channel. It is _not_ decorated as a function on `logging.Logger` (see [here](https://github.com/IBM/alchemy-logging/blob/main/src/python/alog/alog.py#L527))